pkg/sync/grpc/grpc_sync.go

```diff
@@ -0,0 +1,186 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc/credentials/insecure"
+
+	"buf.build/gen/go/open-feature/flagd/grpc/go/sync/v1/syncv1grpc"
+	v1 "buf.build/gen/go/open-feature/flagd/protocolbuffers/go/sync/v1"
+
+	"github.com/open-feature/flagd/pkg/logger"
+	"github.com/open-feature/flagd/pkg/sync"
+	"google.golang.org/grpc"
+)
+
+const (
+	// Prefix for GRPC URL inputs. GRPC does not define a prefix through standard. This prefix helps to differentiate
+	// remote URLs for REST APIs (i.e - HTTP) from GRPC endpoints.
+	Prefix = "grpc://"
+
+	// Connection retry constants
+	// Back off period is calculated with backOffBase ^ #retry-iteration. However, when #retry-iteration count reach
+	// backOffLimit, retry delay fallback to constantBackOffDelay
+	backOffLimit         = 3
+	backOffBase          = 4
+	constantBackOffDelay = 60
+)
+
+type Sync struct {
+	Target     string
+	ProviderID string
+	Logger     *logger.Logger
+}
+
+func (g *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
+	options := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	// initial dial and connection. Failure here must result in a startup failure
+	dial, err := grpc.DialContext(ctx, g.Target, options...)
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error establishing grpc connection: %s", err.Error()))
+		return err
+	}
+
+	serviceClient := syncv1grpc.NewFlagSyncServiceClient(dial)
+	syncClient, err := serviceClient.SyncFlags(ctx, &v1.SyncFlagsRequest{ProviderId: g.ProviderID})
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error calling streaming operation: %s", err.Error()))
+		return err
+	}
+
+	// initial stream listening
+	err = g.handleFlagSync(syncClient, dataSync)
+	if err != nil {
+		g.Logger.Warn(fmt.Sprintf("Error with stream listener: %s", err.Error()))
+	}
+
+	// retry connection establishment
+	for {
+		syncClient, ok := g.connectWithRetry(ctx, options...)
+		if !ok {
+			// We shall exit
+			return nil
+		}
+
+		err = g.handleFlagSync(syncClient, dataSync)
+		if err != nil {
+			g.Logger.Warn(fmt.Sprintf("Error with stream listener: %s", err.Error()))
+			continue
+		}
```

Comment:
I think one interesting thing to consider here is the behavior of flag evaluations when we are in a disconnected state...

Currently, if the gRPC server is down, the evaluations still work, and I think that's probably optimal. Even dynamic evaluations will still be "correct", they are just possibly out of sync since the connection is down. The obvious question is whether or not we think this constitutes and evaluation error... we could remove all the flags for that sync from the state when this connection is lost. I think that would be the wrong choice though (I just want to point it out). The other thing we could do is consider this a "Reason.STALE" evaluation at that point.

Just food for thought @Kavindu-Dodan @james-milligan @beeme1mr @skyerus

Reply:
Interesting, I did not think about the flag evaluation perspective.

I checked the existing implementations (ex:- file, k8s and http) and they seem to react differently to errors. For example, k8s CRD removal doesn't do anything other than logging [1] and http sync performs a return [2] (this is a potential bug ? 🤔 ) . 

I think we need to define this behavior and adopt all sync providers. I am leaning toward the `Reason.STALE` suggestion but this also means we need some advanced `flag store` state updates (detecting non-stale -> stale -> non-stale changes per sync source). But the removal of flags when the sync provider has errors in my view is wrong.  

[1] - https://github.com/open-feature/flagd/blob/main/pkg/sync/kubernetes/kubernetes_sync.go#L73
[2] - https://github.com/open-feature/flagd/blob/main/pkg/sync/http/http_sync.go#L53

Reply:
I think we can make this a separate concern, and we can talk there about the other sync providers as well.

Reply:
I created #400 to follow up on this.